### ISSUE TITLE
cartesian_to_irc produces  180 degree dihedrals with random sign

### ISSUE THREAD

**Issue by AlexB67** [RMeli/irc#52]
Hello, I hope I can explain this right, but I observed the issue as described in the title. Take ethane as an example. it produces the following coordinates:

```
New internal
  (redundant) coordinates: 28
*************************************
  7 bonds / Å
*************************************
  (   0,   1)        1.4613
  (   0,   2)        1.0941
  (   1,   3)        1.0941
  (   0,   4)        1.0941
  (   1,   5)        1.0941
  (   0,   6)        1.0941
  (   1,   7)        1.0941
*************************************
  12 angles / °:
*************************************
  (   1,   0,   2)  111.84
  (   0,   1,   3)  111.84
  (   1,   0,   4)  111.84
  (   2,   0,   4)  107.00
  (   0,   1,   5)  111.84
  (   3,   1,   5)  107.00
  (   1,   0,   6)  111.84
  (   2,   0,   6)  107.00
  (   4,   0,   6)  107.00
  (   0,   1,   7)  111.84
  (   3,   1,   7)  107.00
  (   5,   1,   7)  107.00
*************************************
  9 dihedrals / °
*************************************
  (   2,   0,   1,   3)   -60.00
  (   3,   1,   0,   4)    60.00
  (   2,   0,   1,   5)   180.00
  (   4,   0,   1,   5)   -60.00
  (   3,   1,   0,   6)   -180.00
  (   5,   1,   0,   6)    60.00
  (   2,   0,   1,   7)    60.00
  (   4,   0,   1,   7)  -180.00
  (   6,   0,   1,   7)   -60.00
*************************************

```
If you run cartesian_to_iirc successive times the outcome of the 180 dihedral angles flip sign.  it may give 
```
 (   2,   0,   1,   5)   180.00
 (   3,   1,   0,   6)  -180.00
 (   4,   0,   1,   7)  -180.00
```

or sometimes

```
(   2,   0,   1,   5)  -180.00
(   3,   1,   0,   6)  180.00
(   4,   0,   1,   7)  180.00
```

but only for the 180 degree case. 

As you can imagine this can pay havoc with geometry optimization because dihedrals that have flipped 360 degrees during a geometry optimization step can give rogue values for dq, 

one can get dq = 2 pi when it should have been dq = 0, or something very small.

Yesterday, I put a plaster in my own code to detect this flip, so I could work around the issue. Of course, it would be nice to see what the issue in libirc is.

Thus far, I've made the following change; It works in all my tests so far, but not knowing your code base very well I don't if there could be unwanted consequences

In constsnts.h I added

```
// dihedrals near 2 pi
constexpr double dihedral_near_180_tol{1e-6};

```
and in connectivity.h I changed the following lines, around 440 thereabouts

```
// Compute dihedral angle in radians (in the interval [-pi,pi])
  double angle{std::atan2(y, x)};
  
  angle = (std::fabs(std::fabs(angle) - tools::constants::pi) < tools::constants::dihedral_near_180_tol) 
        ? std::fabs(angle) : angle;

  return angle;
```
Noting that the sign flip occurs because x is sometimes -0.0 and other times +0.0. hence, atan flips the overall sign.

What do you think ?

Best regards.

**Comment by peterbygrave** [comment]
I think you should be determining all differences (including "error" column) in dihedral angles in the `[-180, 180)` domain. If you don't do this then `-179` to `179` is a `358` change rather than the more logical `2` change. The thresholding only helps for that small window, you still have the same problem at `-180+thr+epsilon`.

**Comment by RMeli** [comment]
> The reproducibility of the values is a little more concerning, one should be able to write a test just in the IRC library that confirms there is a problem.

Agreed. I added a test with ethane where `cartesian_to_irc` is called multiple (1000) times and the results are consistent on my end.

@AlexB67 could you please provide a bit more information on the system and the functions you are calling so that I can try to reproduce the issue? If I use the following cartesian coordinates (in angstrom)
```
8

C -0.7560    0.0000    0.0000
C  0.7560    0.0000    0.0000
H -1.1404    0.6586    0.7845
H -1.1404    0.3501   -0.9626
H -1.1405   -1.0087    0.1781
H  1.1404   -0.3501    0.9626
H  1.1405    1.0087   -0.1781
H  1.1404   -0.6586   -0.7845
```
I consistently get the following internal coordinates for the dihedral angles (in radians):
```
 1.0472
 3.1416
-1.0472
-1.0472
 1.0472
 3.1416
 3.1416
-1.0472
 1.0472
```

**Comment by AlexB67** [comment]
Hi,

Thanks for replying. Leave it with me. I would hate to discover it;s something on my end I am doing. I understanding the code base bit by bit. When I have a bit of time, it'll require a strong coffee, and make sure I understand it better.

Yes, I am talking about successive back and forth. Please don't spend any time on my behalf on it until I have a better understanding.   

Best ...

**Comment by AlexB67** [comment]
> I think you should be determining all differences (including "error" column) in dihedral angles in the `[-180, 180)` domain. If you don't do this then `-179` to `179` is a `358` change rather than the more logical `2` change. The thresholding only helps for that small window, you still have the same problem at `-180+thr+epsilon`.

Yes, I had considered that It was a kind of temporary thing to experiment if it worked around the near 180 case in my geometry optimization, so for now I changed nothing in libirc

I thought i had a bug in my BFGS step, but I don't, that's how I came to discover the whole thing.  All the updates for RFO, ,BFGS Hessian matched the output of Psi4 once i just kept track of the sign flips in my own code, but it's probably not 100% full proof.

**Comment by AlexB67** [comment]
> The reproducibility of the values is a little more concerning, one should be able to write a test just in the IRC library that confirms there is a problem.

I'll also do this.  i'll keep you posted.

Thx.

**Comment by RMeli** [comment]
@AlexB67 I added a small test for consecutive forward and backward transformations (this is where the error appears if I understand correctly?) as follows:
```cpp
    const vec irc_0 = cartesian_to_irc<vec3, vec>(x_c, B, A, D, LA, OOPB);
    vec irc = irc_0;

    size_t n_irc = irc.size();

    // Small test on reproducibility, related to issue #52
    for(std::size_t n{0}; n < 100; n++){
      auto result = irc_to_cartesian<vec3, vec, mat>(
          irc, linalg::zeros<vec>(n_irc), x_c, B, A, D, LA, OOPB);

      REQUIRE(result.converged);

      const vec x_c_new = result.x_c;
      irc = cartesian_to_irc<vec3, vec>(x_c_new, B, A, D, LA, OOPB);

      // Check IRC are the same as the first one
      REQUIRE(irc.size() == irc_0.size());
      for(size_t i{0}; i < n_irc; i++){
        CHECK(irc(i) == Approx(irc_0(i)));
      }
    }
```
Unfortunately, it does not seem to reproduce the problem (internal coordinates always match the first conversion). 

Please correct me if I am misunderstanding the problem, but without more details on how to reproduce it in the library it will be tricky to pinpoint the issue.
